se2_planning/include/se2_planning/GridMapLazyStateValidator.hpp

```diff
@@ -23,32 +25,56 @@ class GridMapLazyStateValidator : public GridMapStateValidator {
 
   void setSeed(int value);
   int getSeed() const;
 
+  void setStateValidityCheckingMethod(StateValidityCheckingMethod value);
+  StateValidityCheckingMethod getStateValidityCheckingMethod() const;
+
+  void setStateValidityThreshold(double value);
+  double getStateValidityThreshold() const;
+
+  void setUnsafeStateValidityThreshold(double value);
+  double getUnsafeStateValidityThreshold() const;
+
+  void setMaxNumberOfUnsafeCells(int value);
+  int getMaxNumberOfUnsafeCells() const;
+
  private:
   std::vector<Vertex> nominalFootprintPoints_;
   bool isInitializeCalled_ = false;
   bool isUseRandomizedStrategy_ = false;
   bool isUseEarlyStoppingHeuristic_ = false;
+  StateValidityCheckingMethod stateValidityCheckingMethod_ = StateValidityCheckingMethod::COLLISION;
+  double minStateValidityThreshold_ = 0.5;
+  double unsafeStateValidityThreshold_ = 0.5;
+  int maxNumberOfUnsafeCells_ = 0;
   int seed_ = 0;
 };
 
 void computeFootprintPoints(const grid_map::GridMap& gridMap, const RobotFootprint& footprint, std::vector<Vertex>* footprintPoints);
-bool isInCollision(const SE2state& state, const std::vector<Vertex>& nominalFootprintPoints, const grid_map::GridMap& gridMap,
-                   const std::string& obstacleLayer);
+bool isInCollision(const SE2state& state, const std::vector<Vertex>& footprint, const grid_map::GridMap& gridMap,
+                   const std::string& obstacleLayer, const double collisionThreshold);
+// bool isTraversable(const SE2state& state, const std::vector<Vertex>& footprint, const grid_map::GridMap& gridMap,
+//                   const std::string& traversabilityLayer, const double traversabilityThreshold);
+bool isTraversableIterator(const SE2state& state, const RobotFootprint& footprint, const grid_map::GridMap& gridMap,
+                           const std::string& traversabilityLayer, const double traversabilityThreshold);
+bool isTraversableRobustIterator(const SE2state& state, const RobotFootprint& footprint, const grid_map::GridMap& gridMap,
+                                 const std::string& traversabilityLayer, const double minTraversabilityThreshold,
+                                 const double unsafeTraversabilityThreshold, const int maxNumberOfUnsafeCells);
 void addExtraPointsForEarlyStopping(const RobotFootprint& footprint, std::vector<Vertex>* points, int seed);
-std::unique_ptr<GridMapLazyStateValidator> createGridMapLazyStateValidator(const grid_map::GridMap& gridMap,
-                                                                           const RobotFootprint& footprint,
-                                                                           const std::string& obstacleLayer);
+std::unique_ptr<GridMapLazyStateValidator> createGridMapLazyStateValidator(
+    const grid_map::GridMap& gridMap, const RobotFootprint& footprint, const std::string& obstacleLayer,
+    const StateValidityCheckingMethod& stateValidityCheckingMethod, const double stateValidityThreshold,
```

Comment:
do all methods require all of these parameters? e.g. if I just use the obstacle method?

Reply:
No, you are right, this should probably be organized differently. The `stateValidityThreshold` is used by all three functions, the `unsafeTraversabilityThreshold` and `maxNumberOfUnsafeCells` is only used by the robust traversability function.

Reply:
yes, I think we should remove the functions that are not used from the function calls.